### PR TITLE
 📨 Generic transaction discord messages

### PR DESCRIPTION
**Pull request by bdeneux**
#### 📝 Description
This PR is an improvement of the transaction handler to avoid send the same message on discord on different transaction type and remove the responsibility of the transaction handler to manage action after transaction result.

#### 🏗 Architecture

- `SendMessage` actor message has been modified, it's now hold a generic of type `Message` trait. Message trait is responsible of call the discord library to send the message. 
- `TxHandler` now hold an actor address that implement the `Handler<TxResult>` trait. The `TxResult` is an actor message that could be implemented on all actor that could handle transaction result like Faucet actor. Faucet actor implement it and on transaction result, a discord message is send. **This is no longer the TxHandler that is responsible of the transaction result action.** 
- `TransactionMessage` trait allow to generate the message to send to discord actor based on the transaction result and subscribers. 

#### 🔝 Improvement 

- To avoid cyclic dependency, between the `Faucet` and the `TxHandler`, we should register the response handler (Faucet) after `TxHandler` and `Faucet` initialization. To do that and avoid also mutable reference, the TxHandler handle a new message `RegisterResponseHandler` that will be called to register the Actor who will receive the TxResult message. 

- Also the TxHandler hold two generic (Msg and Handler<TxResult>), because the response handler is registered after the TxHandler initialization, It would be nicer to remove the Handler<TxResult> generic and allowing optional response handler. 
```rust
pub struct TxHandler<T>
where
    T: Msg + Unpin,
{

    ....

    /// Hold address of actor that will receive message when a transaction has been broadcasted.
    response_handler: Option<Addr<dyn Handler<TxResult>>,
}
```
By using the `Handler` trait, we should use the `dyn` keyword. But that cause trouble on all project 😞.

### CODE REVIEW

File: src/cosmos/faucet/discord_message.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+use serenity::async_trait;
+use serenity::http::Http;
+use serenity::model::channel::Message;
+use serenity::model::id::ChannelId;
+use serenity::model::user::User;
+use serenity::prelude::Mentionable;
+use serenity::Error;
+use tracing::{error, info};
+
+/// The faucet respons emessage when transaction successful
```
Comment (ad2ien) [inline src/cosmos/faucet/discord_message.rs:16]:
```suggestion
/// Faucet transaction response
```
(there's a message even if the transaction failed)

File: src/discord/discord_client/message.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+//! Contains definition of a discord message
+
+use serenity::async_trait;
+use serenity::http::Http;
+use serenity::model::channel::Message;
+use serenity::Error;
+
+/// Trait representing a disorcd message
```
Comment (ad2ien) [inline src/discord/discord_client/message.rs:8]:
```suggestion
/// Trait representing a discord message
```

File: src/discord/discord_client/message.rs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+#[async_trait]
+pub trait DiscordMessage {
+    /// Return the channel id relative to send discord message.
```
Comment (ad2ien) [inline src/discord/discord_client/message.rs:11]:
```suggestion
    /// Return the channel id to send discord messages to.
```

**Review by ad2ien — APPROVED** [review]
LGTM, I just pointed out some possible rephrasings 😉

**Comment by ccamel** [comment]
I put it in `draft`, for further analysis with @amimart.

**Review by ccamel — APPROVED** [review]
Ler's go! :rocket:

**Pull request merged**